fabtests/common/shared.c

```diff
@@ -349,6 +349,64 @@ void ft_free_bit_combo(uint64_t *combo)
 	free(combo);
 }
 
+struct ft_context *ft_get_ctx_index(struct ft_context **ctx_arr, int i, size_t mr_size)
+{
+	if (!ctx_arr)
+		return NULL;
+
+	return (struct ft_context *) ((char *) (*ctx_arr) + (i * mr_size));
+}
```

Comment:
I would keep the ft_context array separate from the data transfer buffers.  That is, have an array of ft_context structures.  Allocate the data buffers separately, the way it is today.  If the data buffer can't be re-used (e.g. data verification is enabled), allocate a larger data buffer.